File: Python/Binary_Search.py
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 18 19:50:22 2017

@author: Stehvin
"""

def binarySearch(arr, val):
    """Searches a sorted "array" (list) for a value using a 
    divide-and-conquer approach. At each step, one half of the array is 
    thrown out, based on whether the search value is less than or greater
    than the middle value.
    """
    # initialize starting values
    high = len(arr) - 1
    low = 0
    mid = high//2   
    
    # loop until the value is found
    while arr[mid] != val:
        
        # if the search value is less than the middle value, throw out
        # the right side, otherwise, throw out the left side
        if arr[mid] > val:
            high = mid - 1
        else:
            low = mid + 1
        
        # return None if value could not be found in array
        if high < low:
            return None
        
        # calculate new middle index        
        mid = (high + low)//2
    return mid

File: C#/Binary_Search.cs
using System;

namespace Search
{
	class Program
	{
		static int Main()
		{
            // initialize sorted array and value to search for
            int[] arr = { 2, 3, 4, 13, 24, 50, 56, 67, 70, 245};
            int val = 56;

            // search for value, output index if found, output "-1" if
            // not found
            int index = binarySearch(arr, 0, arr.Length - 1, val);
            Console.WriteLine("{0}", index);
			return 0;
		}

        // search for a value in an array using binary search
        static int binarySearch(int[] arr, int low, int high, int val)
        {
            // return -1 if value could not be found
            if (low > high)
                return -1;
            
            // calculate middle index
            int mid = (high + low) / 2;

            // If value is lower or higher than the middle element, change the
            // index boundaries and repeat the search. If value equals the
            // middle index, return the middle index.
            if (val < arr[mid])
                return binarySearch(arr, low, mid - 1, val);
            else if (val > arr[mid])
                return binarySearch(arr, mid + 1, high, val);
            else
                return mid;
        }
	}
}